waspc/data/Generator/templates/server/src/core/auth/middleware.js

```diff
@@ -0,0 +1,86 @@
+  const defaultValidations = [
+    { validates: EMAIL_FIELD, message: 'email must be present', validator: data => !!data.email, always_on: ['create'] },
+    { validates: PASSWORD_FIELD, message: 'password must be present', validator: data => !!data.password, always_on: ['create'] },
+    { validates: PASSWORD_FIELD, message: 'password must be at least 8 characters', validator: data => data.password.length >= 8 },
+    { validates: PASSWORD_FIELD, message: 'password must contain a number', validator: data => /\d/.test(data.password) },
+  ]
+
+  const validations = [
+    ...(args._waspSkipDefaultValidations ? [] : defaultValidations),
+    ...(args._waspCustomValidations || [])
+  ]
+
+  for (const v of validations) {
+    // If this validation must run, or it optionally can run and has the field
+    if ((v.always_on && v.always_on.includes(action)) || data.hasOwnProperty(v.validates)) {
+      if (!v.validator(data)) {
+        throw new AuthError(v.message)
+      }
+    }
+  }
+}
+
+// Ensure strong plaintext password. Must come before hashing middleware.
+// Throws an AuthError on the first validation that fails.
+const registerPasswordValidation = (prismaClient) => {
+  prismaClient.$use(async (params, next) => {
+    if (params.model === '{= userEntityUpper =}') {
+      if (['create', 'update', 'updateMany'].includes(params.action)) {
+        validateUserData(params.args.data, params.args, params.action)
+      } else if (params.action === 'upsert') {
+        validateUserData(params.args.create.data, params.args, 'create')
+        validateUserData(params.args.update.data, params.args, 'update')
+      }
+
+      // Remove from downstream Prisma processing to avoid "Unknown arg" error
+      delete params.args._waspSkipDefaultValidations
+      delete params.args._waspCustomValidations
+    }
+
+    return next(params)
+  })
+}
+
+// Make sure password is always hashed before storing to the database.
+const registerPasswordHashing = (prismaClient) => {
+  prismaClient.$use(async (params, next) => {
+    if (params.model === '{= userEntityUpper =}') {
+      if (['create', 'update', 'updateMany'].includes(params.action)) {
+        if (params.args.data.hasOwnProperty(PASSWORD_FIELD)) {
+          params.args.data[PASSWORD_FIELD] = await hashPassword(params.args.data[PASSWORD_FIELD])
+        }
+      } else if (params.action === 'upsert') {
+        if (params.args.create.data.hasOwnProperty(PASSWORD_FIELD)) {
+          params.args.create.data[PASSWORD_FIELD] =
+            await hashPassword(params.args.create.data[PASSWORD_FIELD])
+        }
+        if (params.args.update.data.hasOwnProperty(PASSWORD_FIELD)) {
+          params.args.update.data[PASSWORD_FIELD] =
+            await hashPassword(params.args.update.data[PASSWORD_FIELD])
+        }
+      }
+    }
+
+    return next(params)
+  })
+}
+
+export const registerAuthMiddleware = (prismaClient) => {
```

Comment:
Nice, this way we make sure the order is correct here, where we are close to the assumption that makes order important.

Reply:
Just to add on this -> I believe you have a comment in the header of registerUserEntityValidation middleware that says it needs to go before password is hashed. I would actually consider moving that comment here, since here is the location where somebody can potentially mess it up + here we have higher-level look where we are looking at both middlewares and deciding which one goes first.

Reply:
Good catch, thanks! This is addressed here: https://github.com/wasp-lang/wasp/pull/376/commits/a72eb8d40f25d34d2c882cf72fe6fca28f0a2ca3